### PR TITLE
Only mark copying of assets to the share folder as done when we know it was successful

### PR DESCRIPTION
**Pull request by nirvn**

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e9ad0de858387da4ec0588f9d9147d51b7f5ab72:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-0b6a54a3aa13d03157e9a0cea28220e1cca3aac1-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-0b6a54a3aa13d03157e9a0cea28220e1cca3aac1-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-0b6a54a3aa13d03157e9a0cea28220e1cca3aac1-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-0b6a54a3aa13d03157e9a0cea28220e1cca3aac1-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 1531fc540df12a005ab14212f3ff9a7789e44f89:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-da61b9190ed7738988bc2a484d683fee9c0dff98-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-da61b9190ed7738988bc2a484d683fee9c0dff98-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-da61b9190ed7738988bc2a484d683fee9c0dff98-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-da61b9190ed7738988bc2a484d683fee9c0dff98-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for d78eecf4f2a271ffca3e055b82895e86ef2191a3:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-052bd35f334065208e3c4e4aef324d95be5793ca-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-052bd35f334065208e3c4e4aef324d95be5793ca-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-052bd35f334065208e3c4e4aef324d95be5793ca-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-052bd35f334065208e3c4e4aef324d95be5793ca-x86.apk)

**Comment by nirvn** [comment]
@m-kuhn , alright, I did a fair amount of checking on the GIT_REV variable, the problem doesn't seem to be there, we get proper values across packages (I've checked assets of both 1.9.6 and 1.10.0, those have proper GIT_REV values and are different).

The one scenario I think we could be hitting randomly would be a failure to actually delete or copy files into share (or an early thread termination outside our control). We currently don't handle those scenarios since once the copying thread is over (without knowing whether it was successful or not) we write the new GIT_REV value within the gitRev file.

To cover those scenarios, I'm proposing that we move the writing of .gitRev file _only_ when we're sure the copying process has been successful. This will insure that we never leave someone with a broken /share folder (either with missing files or extra files that haven't been removed from a previous installation).

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7c67a9b5df63a3bf13dbb2db754cedbe3e8d8453:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-c77071159de8a531b03e6a1b6bc9bed5a9a73253-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-c77071159de8a531b03e6a1b6bc9bed5a9a73253-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-c77071159de8a531b03e6a1b6bc9bed5a9a73253-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-c77071159de8a531b03e6a1b6bc9bed5a9a73253-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for aebef8e50a29d1d507d769f873ff65aab0dfb52c:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e78bdff9206400e1cd6facfc171ab4f8c0035c62-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e78bdff9206400e1cd6facfc171ab4f8c0035c62-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e78bdff9206400e1cd6facfc171ab4f8c0035c62-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e78bdff9206400e1cd6facfc171ab4f8c0035c62-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5433f28ea7c98383b2620a7aad2529d0dca71cdf:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a3c95a2b1154bf75db2215c80046d8626e4b45ad-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a3c95a2b1154bf75db2215c80046d8626e4b45ad-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a3c95a2b1154bf75db2215c80046d8626e4b45ad-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a3c95a2b1154bf75db2215c80046d8626e4b45ad-x86.apk)

File: src/core/utils/fileutils.cpp [hunk 1]
```diff
@@ -58,9 +58,18 @@ bool FileUtils::copyRecursively( const QString &sourceFolder, const QString &des
   // Remove the destination folder and its content if it already exists
   if ( wipeDestFolder )
   {
+    qDebug() << "removing " + destFolder;
```
Comment (m-kuhn) [inline src/core/utils/fileutils.cpp:61]: Happy to leave these debug statements if they are more verbose to non involved readers

Comment (nirvn) [reply src/core/utils/fileutils.cpp:61]: @m-kuhn , I was thinking to at least leave one debug call along the lines of "Different git revision detected, copying data to share directory" so we can get some hint in logcat of users in the future.

Comment (m-kuhn) [reply src/core/utils/fileutils.cpp:61]: And a `%1 files copied to %2` at the end would also be nice

Comment (m-kuhn) [reply src/core/utils/fileutils.cpp:61]: ... I'm thinking about moving this (at least partially) to all platforms. The package data will be r/o in *every* modern well maintained system (root only, admin only, in a :/assets file system ...) so we need to copy at least demo_projects to somewhere else.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 75323f1af36a401827b6af294e49b304b7d59440:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-cda924cc0b805a6080f19e316f6d823bd156c18f-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-cda924cc0b805a6080f19e316f6d823bd156c18f-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-cda924cc0b805a6080f19e316f6d823bd156c18f-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-cda924cc0b805a6080f19e316f6d823bd156c18f-x86.apk)

**Comment by nirvn** [comment]
@m-kuhn , alright, cleaned up the commit and left two debug calls in.

File: src/core/platforms/android/androidplatformutilities.cpp [hunk 2]
```diff
@@ -49,37 +49,50 @@ class FileCopyThread : public QThread
     Q_OBJECT
 
   public:
-    FileCopyThread( const QString &source, const QString &destination, Feedback *feedback )
+    FileCopyThread( const QString &source, const QString &appDataLocation, const QByteArray &gitRev, Feedback *feedback )
```
Comment (m-kuhn) [inline src/core/platforms/android/androidplatformutilities.cpp:52]: Can we keep this generic (to allow copying any dir anywhere?

Comment (nirvn) [reply src/core/platforms/android/androidplatformutilities.cpp:52]: @m-kuhn , how would you handle gitrev writing?

Comment (m-kuhn) [reply src/core/platforms/android/androidplatformutilities.cpp:52]: I.e. adding gitRev (or "versionTag") is ok, making it specific for `appDataLocation` is the part that will make it trickier to separate demo_projects (need to be copied to a writable location) and other data (that only needs to be copied on android where we have a location that's not accessible for gdal, proj etc).

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 57219444fd074eab5af0161d2b573558410bd5c6:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e69ac2e61e6be21e613732eb75a3ceab5801479e-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e69ac2e61e6be21e613732eb75a3ceab5801479e-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e69ac2e61e6be21e613732eb75a3ceab5801479e-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-e69ac2e61e6be21e613732eb75a3ceab5801479e-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 606c546ad44df932bc8555bef49bba23e0281088:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2d9554920c2809a6a41b9eb84322cc60bd84f528-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2d9554920c2809a6a41b9eb84322cc60bd84f528-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2d9554920c2809a6a41b9eb84322cc60bd84f528-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2d9554920c2809a6a41b9eb84322cc60bd84f528-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7343609f0c282c302d5191dd8efc0c8d544fef6d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-56f1574a343cf1e94244e2bdd0a43c94c797b0e7-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-56f1574a343cf1e94244e2bdd0a43c94c797b0e7-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-56f1574a343cf1e94244e2bdd0a43c94c797b0e7-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-56f1574a343cf1e94244e2bdd0a43c94c797b0e7-x86.apk)

File: src/core/platforms/android/androidplatformutilities.cpp [hunk 3]
```diff
@@ -72,14 +73,17 @@ class FileCopyThread : public QThread
...
+
   QFile gitRevFile( appDataLocation + QStringLiteral( "/gitRev" ) );
   gitRevFile.open( QIODevice::ReadOnly );
-  QByteArray appGitRev = getIntentExtra( "GIT_REV" ).toLocal8Bit();
   QByteArray localGitRev = gitRevFile.readAll();
+  gitRevFile.close();
+  QByteArray appGitRev = getIntentExtra( "GIT_REV" ).toLocal8Bit();
   if ( localGitRev != appGitRev )
   {
+    qDebug() << "Different build git revision detected";
```
Comment (m-kuhn) [inline src/core/platforms/android/androidplatformutilities.cpp:86]: Should we include the (first chars of) the git revs in here?

File: src/core/utils/fileutils.cpp [hunk 4]
```diff
@@ -60,7 +59,11 @@ bool FileUtils::copyRecursively( const QString &sourceFolder, const QString &des
   {
     QDir destDir( destFolder );
     if ( destDir.exists() )
-      destDir.removeRecursively();
+    {
+      bool success = destDir.removeRecursively();
+      if ( !success )
+        return false;
```
Comment (m-kuhn) [inline src/core/utils/fileutils.cpp:65]: If we assume this happens, let's add an explicit debug msg here, with reason if possible.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for aa98b08e2384cb93802436e358feb61067397a39:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a8c277418592e9d5918df4009869322168ea6a9a-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a8c277418592e9d5918df4009869322168ea6a9a-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a8c277418592e9d5918df4009869322168ea6a9a-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-a8c277418592e9d5918df4009869322168ea6a9a-x86.apk)

File: src/core/platforms/android/androidplatformutilities.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/platforms/android/androidplatformutilities.cpp:86]: Done

File: src/core/utils/fileutils.cpp (hunk 4 above)
Comment (nirvn) [reply src/core/utils/fileutils.cpp:65]: Done

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 9305acc3a2ad98699cdf58a90b36b931b8ebd06c:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-b6a7ded6e331f7d8a51bd4913a1b46e6912089ea-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-b6a7ded6e331f7d8a51bd4913a1b46e6912089ea-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-b6a7ded6e331f7d8a51bd4913a1b46e6912089ea-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-b6a7ded6e331f7d8a51bd4913a1b46e6912089ea-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 4cdb03c0f75537dc772b6ea47f03566375123806:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2f2c6598817b83a8a56a0c1a27c9144f33e2c030-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2f2c6598817b83a8a56a0c1a27c9144f33e2c030-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2f2c6598817b83a8a56a0c1a27c9144f33e2c030-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2264-2f2c6598817b83a8a56a0c1a27c9144f33e2c030-x86.apk)

**Comment by nirvn** [comment]
@m-kuhn , I think this is ready for merging + backporting.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**